feat(eth-staking): implement staking logic

StanCheb opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

<!--- Describe your changes in detail -->
The staking action logic is implemented. 
The "Stake" transactions list is added to the "Staking" tab.
Fee calculations are corrected.

Tx presented in screenshots: https://goerli1.trezor.io/tx/0xf8c31eb522b200c7d33d628e9b4723072313f572023376dbe947f3c356d0b204

NB! Do not test with real ETH, since unstaking and claiming are not implemented yet. Goerli has only partial support in the wallet SDK. It's recommended to test with Holesky, unless more edge cases with errors are the subject of testing.

## Related Issue

Resolve <!--- link the issue here -->
https://github.com/trezor/trezor-suite/issues/9267
https://github.com/trezor/trezor-suite/issues/9257
https://github.com/trezor/trezor-suite/issues/9255
https://github.com/trezor/trezor-suite/issues/9220

## Screenshots:
<img width="1497" alt="Знімок екрана 2024-01-11 о 16 06 30" src="https://github.com/everstake/trezor-suite/assets/42133844/ab789f08-efef-4662-9585-d1a3c51cf2b1">
<img width="1497" alt="Знімок екрана 2024-01-11 о 16 06 41" src="https://github.com/everstake/trezor-suite/assets/42133844/38a53833-75cc-4024-a03d-425723cd290e">
<img width="1497" alt="Знімок екрана 2024-01-11 о 16 06 49" src="https://github.com/everstake/trezor-suite/assets/42133844/709b6987-517c-4767-b22e-5ea602c3b897">
<img width="1497" alt="Знімок екрана 2024-01-11 о 16 07 01" src="https://github.com/everstake/trezor-suite/assets/42133844/bc621536-8234-42c2-8ed3-e35da9cea286">
<img width="1497" alt="Знімок екрана 2024-01-11 о 16 07 32" src="https://github.com/everstake/trezor-suite/assets/42133844/6eff3790-77ab-4f6c-8c5d-0d80c4f8beeb">

StanCheb committed:
feat(eth-staking): implement staking logic

tomasklim commented on suite-common/suite-utils/src/stake.ts:
```diff
@@ -0,0 +1,6 @@
+export const isStakeTx = (txName: string | undefined) => {
+    if (!txName) return false;
+
+    const stakeTxNames = ['Stake', 'Unstake', 'Claim'];
```
It would be better to use `Bytes Signatures` here.

`0x3a29dbae` for staking
`0x33986ffa` for claim
...

`tx.ethereumSpecific?.parsedData?.methodId`

StanCheb replied:
https://github.com/everstake/trezor-suite/pull/7/commits/93e57e6503e1b0ec2120169aa444c7dcc621a7b2

tomasklim commented on packages/suite/src/constants/suite/ethStaking.ts:
```diff
@@ -3,6 +3,8 @@ import BigNumber from 'bignumber.js';
 export const MIN_ETH_AMOUNT_FOR_STAKING = new BigNumber(0.1);
 export const MIN_ETH_FOR_WITHDRAWALS = new BigNumber(0.03);
 export const MIN_ETH_BALANCE_FOR_STAKING = MIN_ETH_AMOUNT_FOR_STAKING.plus(MIN_ETH_FOR_WITHDRAWALS);
+// Source is used for wallet sdk txs to define what app requests a tx.
```
I do not understand the comment, could you please improve it? I would add link to docs. However, it is not explained there either 

https://wallet-sdk.everstake.one/guide/ethereum/stake.html

StanCheb replied:
https://github.com/everstake/trezor-suite/pull/7/commits/93e57e6503e1b0ec2120169aa444c7dcc621a7b2

tomasklim commented on packages/suite/package.json:
```diff
@@ -19,6 +19,7 @@
         "test-unit:watch": "jest -o --watch"
     },
     "dependencies": {
+        "@everstake/wallet-sdk": "^0.3.37",
```
This library should be moved to `@trezor/connect`. 
All the crypto logic (ethereum, cardano, solana,...) should be located there

tomasklim replied:
Let's do this in second iteration as we discussed

tomasklim commented on packages/suite/src/utils/suite/stake.ts:
```diff
@@ -30,3 +37,69 @@ export const getStakeFormsDefaultValues = ({
+          errorMessage: string;
+      };
+
+export const prepareStakeEthTx = async ({
```
I would move this part of file to `@trezor/connect` together with the everstake sdk

=> `src/api/ethereum`

tomasklim commented on packages/suite/src/actions/wallet/stake/stakeFormEthereumActions.ts:
```diff
@@ -98,37 +100,35 @@ export const composeTransaction =
+        try {
+            Ethereum.selectNetwork(getEthNetworkForWalletSdk(account.symbol));
+            // TODO: Implement tx type switcher
+            txData = await Ethereum.stake(account.descriptor, amount, WALLET_SDK_SOURCE);
```
Move to Connect

tomasklim commented:
Unfortunately, I wasn't able to stake

1. If `10%` is less than the required amount, I would disable the 10% button with some tooltip (applies also to others)
<img width="384" alt="Screenshot 2024-01-15 at 12 10 57" src="https://github.com/everstake/trezor-suite/assets/33235762/81e83ebf-e4b6-4f1f-bb90-a8efeb8d4ea2">

2. If I click on `max` it sets max minus 0.03 in the input but in the review modal, there is max amount and not the max - 0.03
<img width="750" alt="Screenshot 2024-01-15 at 22 00 52" src="https://github.com/everstake/trezor-suite/assets/33235762/72f18efb-95dc-4554-a712-b73aa58a5700">

And if I confirm everything on Trezor and start to stake, I get this error
<img width="893" alt="Screenshot 2024-01-15 at 12 24 45" src="https://github.com/everstake/trezor-suite/assets/33235762/7b2804b2-664c-446a-a5ec-8b719e1bc84d">

3. When I enter whatever amount except max, I cant get to the next step. `Confirm & stake` does not work for me
<img width="616" alt="Screenshot 2024-01-15 at 21 58 36" src="https://github.com/everstake/trezor-suite/assets/33235762/4d20401d-47d0-450b-8e6c-9855217d506a">

4. If I click on `max` and then I change the value, it is not changed in the review modal

https://github.com/everstake/trezor-suite/assets/33235762/7efe803b-53c4-4e31-9c05-69914d284303

5. Staking does not work in desktop app as there is some request to your backend which is not whitelisted
<img width="1070" alt="Screenshot 2024-01-15 at 21 59 03" src="https://github.com/everstake/trezor-suite/assets/33235762/0b4b286b-0b3e-4f69-abed-6c79529ffadc">

<img width="720" alt="Screenshot 2024-01-15 at 21 58 55" src="https://github.com/everstake/trezor-suite/assets/33235762/1a2ad6b8-526b-41ee-b3c2-5328ec138e55">

6. Disable or hide the fiat input for testnet as we do not know rates (it is hidden in send form). Util `isTestnet`
<img width="434" alt="Screenshot 2024-01-15 at 12 27 42" src="https://github.com/everstake/trezor-suite/assets/33235762/f467fee5-0504-4886-a283-2acb510d5898">

7. On Sepolia, I am getting this error (maybe it is not supported)

<img width="1189" alt="Screenshot 2024-01-15 at 12 12 18" src="https://github.com/everstake/trezor-suite/assets/33235762/ba0bc923-2471-4fd8-85cb-44eb2b42da1f">

8. Hide fiat value for testnets, we have util `isTestnet`

<img width="815" alt="Screenshot 2024-01-15 at 22 14 57" src="https://github.com/everstake/trezor-suite/assets/33235762/60dc7fa3-a343-4f5f-a951-5eda6a940a11">

9. If I get back to form from the review modal, I do not see clear button
<img width="431" alt="Screenshot 2024-01-15 at 22 19 13" src="https://github.com/everstake/trezor-suite/assets/33235762/9dcd6d0e-4648-4674-ba72-054aa3100b96">

StanCheb commented:
> Unfortunately, I wasn't able to stake
> 
> 1. If `10%` is less than the required amount, I would disable the 10% button with some tooltip (applies also to others)
> 
> <img alt="Screenshot 2024-01-15 at 12 10 57" width="384" src="https://private-user-images.githubusercontent.com/33235762/296928401-81e83ebf-e4b6-4f1f-bb90-a8efeb8d4ea2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5Mjg0MDEtODFlODNlYmYtZTRiNi00ZjFmLWJiOTAtYThlZmViOGQ0ZWEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiM2E3MzQxZTk4ZWRjZTJjOGI5YzU2YjZmOWVkNGZmNjVlMjljOTQ2YzUzMTExZmU5ZGEyNjUxOWIxNmQ0YzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NcJJo9U5p8vRmK9FHEE7v5iRJzoSw_6wQvjgONgCjog">
> 2. If I click on `max` it sets max minus 0.03 in the input but in the review modal, there is max amount and not the max - 0.03
> 
> <img alt="Screenshot 2024-01-15 at 22 00 52" width="750" src="https://private-user-images.githubusercontent.com/33235762/296928550-72f18efb-95dc-4554-a712-b73aa58a5700.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5Mjg1NTAtNzJmMThlZmItOTVkYy00NTU0LWE3MTItYjczYWE1OGE1NzAwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0YTBmM2EwZjc0YWZhZGY5YWI2ZDYzMWMzMzZhNDRiMDZjMTVmMjQ3MDFhYTgyNTFjMGQ4OTJmNjA0NTJmYjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.XNsL3WwJ_1F74rImS9cihD4b2B2F8hZNaDoZ0lYbmJU">
> And if I confirm everything on Trezor and start to stake, I get this error <img alt="Screenshot 2024-01-15 at 12 24 45" width="893" src="https://private-user-images.githubusercontent.com/33235762/296929452-7b2804b2-664c-446a-a5ec-8b719e1bc84d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5Mjk0NTItN2IyODA0YjItNjY0Yy00NDZhLWE1ZWMtOGI3MTllMWJjODRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ5ZjdkNDNmNGI0ZTFmZWViNWYzYzI0MTg0MWUyZjIxMDgyMWM0NWU0ZjUyZjgyMzgwOThiMDYzODMxN2FkOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MQYpPdNgltQei56ZjUwlNHIwkpTdmZCmJt5Doq8QJqk">
> 
> 3. When I enter whatever amount except max, I cant get to the next step. `Confirm & stake` does not work for me
> 
> <img alt="Screenshot 2024-01-15 at 21 58 36" width="616" src="https://private-user-images.githubusercontent.com/33235762/296928987-4d20401d-47d0-450b-8e6c-9855217d506a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5Mjg5ODctNGQyMDQwMWQtNDdkMC00NTBiLThlNmMtOTg1NTIxN2Q1MDZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2MzkxNWYwMjY5ZWI5MjgxMDQ4NDExM2IxZDFkNTk1ODFkNTkzOGJjMDI4N2VjZmJlMDA2NzIyYTlmMTkwZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IEv-giSnOz5kceCIfiQ0yG3z4xMo_NTJtEQ_-5gK0d0">
> 4. If I click on `max` and then I change the value, it is not changed in the review modal
> 
>  Screen.Recording.2024-01-15.at.12.25.56.mov 
> 5. Staking does not work in desktop app as there is some request to your backend which is not whitelisted
> 
> <img alt="Screenshot 2024-01-15 at 21 59 03" width="1070" src="https://private-user-images.githubusercontent.com/33235762/296928878-0b4b286b-0b3e-4f69-abed-6c79529ffadc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5Mjg4NzgtMGI0YjI4NmItMGIzZS00ZjY5LWFiZWQtNmM3OTUyOWZmYWRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxMWZhYTg2MzcwMTNiOGQyMTA0NmE5ODJmOTAxOGFjN2RjNDRkYjA4Y2Y4ZmRiNGQ5MDBjODgzYTliZDMwNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VT5YIuaFuIJvZfNMiVGHHKA-YlHhU-UNj5DJT0T3eNY"> <img alt="Screenshot 2024-01-15 at 21 58 55" width="720" src="https://private-user-images.githubusercontent.com/33235762/296928894-1a2ad6b8-526b-41ee-b3c2-5328ec138e55.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5Mjg4OTQtMWEyYWQ2YjgtNTI2Yi00MWVlLWIzYzItNTMyOGVjMTM4ZTU1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc2YWE1NzQ4NmY0ZGIyZmM2ZjBlMjIyMmRjMzE2ZDkxM2VhMDg4NGQ3YzQwODNlYzBiNjEyYmY4OWNlNmExNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PPr1Z2AqE5fUV8kU0kBdIysWk6LkL3wmAMqXXN_3UdA">
> 6. Disable or hide the fiat input for testnet as we do not know rates (it is hidden in send form). Util `isTestnet`
> 
> <img alt="Screenshot 2024-01-15 at 12 27 42" width="434" src="https://private-user-images.githubusercontent.com/33235762/296929184-f467fee5-0504-4886-a283-2acb510d5898.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5MjkxODQtZjQ2N2ZlZTUtMDUwNC00ODg2LWEyODMtMmFjYjUxMGQ1ODk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzMTYyNTlmYjQ0MmU4YmRiM2JlOWFhMjM2NDNjNzgzZDY4OGI5NmQzNGNjNzY0Mzc2Y2U0YjZjODA3YThiZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tFBW7m0idXP0tnE7-09e-ktb4ZAEJi-BCJFSpoIl8qE">
> 7. On Sepolia, I am getting this error (maybe it is not supported)
> 
> <img alt="Screenshot 2024-01-15 at 12 12 18" width="1189" src="https://private-user-images.githubusercontent.com/33235762/296929606-ba0bc923-2471-4fd8-85cb-44eb2b42da1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5Mjk2MDYtYmEwYmM5MjMtMjQ3MS00ZmQ4LTg1Y2ItNDRlYjJiNDJkYTFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5NDRhMGJkYWQzODY4N2JlZGRjNGVjZmNkNmJjN2NkMjlhZjY0NjkwY2YyZjU4M2I1NGU1NTQ1ZjgzNDE5YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mk4pGy9-JNyfFD8pO0FrQn1E8k94vwy6Eur_NWZ5L3M">
> 8. Hide fiat value for testnets, we have util `isTestnet`
> 
> <img alt="Screenshot 2024-01-15 at 22 14 57" width="815" src="https://private-user-images.githubusercontent.com/33235762/296930205-60dc7fa3-a343-4f5f-a951-5eda6a940a11.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5MzAyMDUtNjBkYzdmYTMtYTM0My00ZjVmLWE5NTEtNWVkYTZhOTQwYTExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNiNDk2YWIxMDk4NDYyYTI2MGU3YTdiYThjYzAwNWI2ZmE1ODk5NGY4OTllZTAwOWY2ZTYxYzMyNjk4YmE5NzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.30ZEQ8730J_SpMpF5nco2Hw_lbjLeSpA1CooIRrw6PQ">
> 9. If I get back to form from the review modal, I do not see clear button
> 
> <img alt="Screenshot 2024-01-15 at 22 19 13" width="431" src="https://private-user-images.githubusercontent.com/33235762/296930878-9dcd6d0e-4648-4674-ba72-054aa3100b96.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU0MjU3MDcsIm5iZiI6MTcwNTQyNTQwNywicGF0aCI6Ii8zMzIzNTc2Mi8yOTY5MzA4NzgtOWRjZDZkMGUtNDY0OC00Njc0LWJhNzItMDU0YWEzMTAwYjk2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE2VDE3MTY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiZDdjNjdmYzdkMWYxY2JiMGY0NWY0Nzc3MWYyYWM0MDBlZGQ4YjRkYzkyOGMzOWU0YjIyYzJlMzgwMGQ3NWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ifdFAQXKpcSuJ3hGZ9XxnDC6CuxXa7xvglZ6fFFeOz8">

1. Done
2. Fixed
3. Not sure what the problem is. The only case when I had a similar behaviour, was when the emulator was disconnected from the app.
4. Fixed
5. Not sure about this one. I thought we were implementing the web part only. Need some clarification with our backend team
6. This is left for testing purposes. It uses actual ETH rates. Will remove that in the very end, when everything is tested.
7. Sepolia is not supported in the SDK. Removed it from available for staking tokens.
8. This is left for testing purposes. It uses actual ETH rates. Will remove that in the very end, when everything is tested.
9. Not sure how you got this state without getting the field «dirty». You also don’t have the fiat input filled, which is strange. Could you provide more details. please? Added a blind fix for now.

https://github.com/everstake/trezor-suite/pull/7/commits/93e57e6503e1b0ec2120169aa444c7dcc621a7b2

tomasklim commented:
1. Nice. I would not disable `Max` button and I would add tooltip on hover to disabled buttons `Minimum amount to stake is 0.1 ETH` 

2. ok

3. see code comment below

4. still there, same reason as number 3, `setMaxOutputId` is not reset because of missing `currentRate`

5. There is almost no difference between web and desktop app, don't worry. The domain just needs to be added to whitelist in `allowedDomains` array in `packages/suite-desktop-core/src/config.ts`. It should also work with tor

6. aha oka, probably same reason number 3

7. ok

8. ok

9. fixed

**10. NEW**

https://github.com/everstake/trezor-suite/assets/33235762/cc1c5bd6-dc38-4e8d-9421-80b3fe3417a6

tomasklim commented on packages/suite/src/hooks/wallet/useStakeEthForm.ts:
```diff
@@ -181,9 +198,11 @@ export const useStakeEthForm = ({ selectedAccount }: UseStakeFormsProps): StakeC
             setIsAmountForWithdrawalWarningShown(false);
             setIsAdviceForWithdrawalWarningShown(false);
-            if (!fiatRates || !fiatRates.current) return;
+            if (!currentRate) return;
```
This condition is a reason for number 3., there is no current rate for testnet coins

tomasklim replied:
https://github.com/everstake/trezor-suite/assets/33235762/f1e2ba76-37c7-47d8-a7a0-ce31fa0a5a33

StanCheb commented:
1. Done.

3 & 4. That’s interesting. Can’t figure out how it’s possible. I use` mapTestnetSymbol() `to get the eth symbol for current rates. Here is this part:

https://github.com/everstake/trezor-suite/pull/7/files#diff-da2957c4b8ba4ab8317bee25a5c27ecefa9d55be5d6da364a287a3489307fdf8R58


https://github.com/everstake/trezor-suite/assets/42133844/ab583c23-475d-4932-b01b-935c7856dc22



Could you, please, check if you have any local changes?

5. Added

10. Another strange case. The message says 0.000003, though it should be 0.03. I wonder how that happened. I didn’t change that constant. The field is supposed to be invalid and empty in this case, since we are unable to calculate max correctly. Is this case reproducible with Goerli or Holesky?

<img width="1511" alt="Знімок екрана 2024-01-17 о 16 49 43" src="https://github.com/everstake/trezor-suite/assets/42133844/d5bd330e-f8aa-427b-8276-3be0e2820ddb">


https://github.com/everstake/trezor-suite/pull/7/commits/b5fddd3335065bdb802ed3d8a2b431b425eed068

StanCheb commented:
UPD

I've managed to get a similar behaviour with empty `currentRate` (points 3 & 4). I get it when there are some problems with websockets. But I'm still not quite sure with your case, since you didn't have any errors in the console regarding websockets. Moreover, I'm still able to proceed with staking even without rates

<img width="1511" alt="Знімок екрана 2024-01-18 о 14 05 47" src="https://github.com/everstake/trezor-suite/assets/42133844/118d95b7-0dd5-402c-bd55-181a6221556e">


https://github.com/everstake/trezor-suite/assets/42133844/12cc6d26-0d6b-445c-9d11-fa49533ac996

tomasklim commented:
> 1. Done.
> 
> 3 & 4. That’s interesting. Can’t figure out how it’s possible. I use`mapTestnetSymbol()`to get the eth symbol for current rates. Here is this part:
> 
> https://github.com/everstake/trezor-suite/pull/7/files#diff-da2957c4b8ba4ab8317bee25a5c27ecefa9d55be5d6da364a287a3489307fdf8R58
> 
>  2024-01-17.16.06.28.mov 
> Could you, please, check if you have any local changes?
> 
> 5. Added
> 6. Another strange case. The message says 0.000003, though it should be 0.03. I wonder how that happened. I didn’t change that constant. The field is supposed to be invalid and empty in this case, since we are unable to calculate max correctly. Is this case reproducible with Goerli or Holesky?
> 
> <img alt="Знімок екрана 2024-01-17 о 16 49 43" width="1511" src="https://private-user-images.githubusercontent.com/42133844/297439431-d5bd330e-f8aa-427b-8276-3be0e2820ddb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2MTkwMzAsIm5iZiI6MTcwNTYxODczMCwicGF0aCI6Ii80MjEzMzg0NC8yOTc0Mzk0MzEtZDViZDMzMGUtZjhhYS00MjdiLTgyNzYtM2JlMGUyODIwZGRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDIyNTg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0NWJhNzFlMzc2MzNhN2FjMTlkMTc5NjhhYTFhOTkxMDE3ZDFmZTE3ZmU0OGU5NWQ3NTdkMDgxYWQ5M2M2MmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OIB1Nrs_A_5pYNgPQiEFbtt1GOwDyRx5IIb3OvSj72k">
> [b5fddd3](https://github.com/everstake/trezor-suite/commit/b5fddd3335065bdb802ed3d8a2b431b425eed068)

Sorry, I had local changes of amount, so number 10 is irrelevant, sorry

tomasklim commented:
> UPD
> 
> I've managed to get a similar behaviour with empty `currentRate` (points 3 & 4). I get it when there are some problems with websockets. But I'm still not quite sure with your case, since you didn't have any errors in the console regarding websockets. Moreover, I'm still able to proceed with staking even without rates
> 
> <img alt="Знімок екрана 2024-01-18 о 14 05 47" width="1511" src="https://private-user-images.githubusercontent.com/42133844/297753895-118d95b7-0dd5-402c-bd55-181a6221556e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU2MTkwMzAsIm5iZiI6MTcwNTYxODczMCwicGF0aCI6Ii80MjEzMzg0NC8yOTc3NTM4OTUtMTE4ZDk1YjctMGRkNS00MDJjLWJkNTUtMTgxYTYyMjE1NTZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTE4VDIyNTg1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0YjY5NmIwOWI5NWE3NzM2NzVjNTU4NWVkMGEzOGViYjA2MzMxOGUxYWQ2NjQ5MTIyMjE2MThmNmEzMDJjZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2_3XZJ5H6WVa8TH1BZf9UYHson9aZ7qDBD5LIYzJtLc">
>  2024-01-18.14.20.54.mov

Regarding this, we are having a DDoS attack our ETH backends, so maybe that's the issue with errors. Fiat rates are working for me now 😀

tomasklim commented on packages/suite-desktop-core/src/config.ts:
```diff
@@ -25,6 +25,9 @@ export const allowedDomains = [
     'blockfrost.dev',
+    'eth-goerli.public.blastapi.io',
+    'ethereum-holesky.publicnode.com',
+    'mainnet.infura.io/v3/f583d4f04d384b9e8c59a7ff1c9f68f1',
```
```suggestion
    'mainnet.infura.io',
```

tomasklim replied:
_(As we discussed, those backends are temporary. I raised a question if we can provide RPC using our blockbook, if needed)_

StanCheb replied:
https://github.com/everstake/trezor-suite/pull/7/commits/6f35bf0e58d70245d04b4ced2b5d2af3d7bf2558

tomasklim commented:
Two last things, see provided video

1. If balance is bigger than `0.1` but the max is not enough (0.13), an error occures
2. if i insert amount which is more than `amount + fee`. It does not show the fee and it allows me to continue. However, continue button in the confirmation modal does nothing
https://github.com/everstake/trezor-suite/assets/33235762/fa845d78-ecef-4465-a723-f1c9923db2b8

As it is kinda edge case, if you want to solve it later, I will approve it after change of infura ⬇️

StanCheb commented:
> Two last things, see provided video
> 
> 1. If balance is bigger than `0.1` but the max is not enough (0.13), an error occures
> 2. if i insert amount which is more than `amount + fee`. It does not show the fee and it allows me to continue. However, continue button in the confirmation modal does nothing
>    https://github.com/everstake/trezor-suite/assets/33235762/fa845d78-ecef-4465-a723-f1c9923db2b8
> 
> As it is kinda edge case, if you want to solve it later, I will approve it after change of infura ⬇️

https://github.com/everstake/trezor-suite/pull/7/commits/6f35bf0e58d70245d04b4ced2b5d2af3d7bf2558

1. Yep, you’re right. I try to compose a tx with the right amount which in the case of clicking max is `balance - 0.03` at this stage. SDK gets the value less than 0.1 and throws an error. As a result we cannot calculate the gas limit and correct max amount, therefore there is nothing to set in the field. That’s why I disabled the max button when` balance - 0.3 < 0.13` before last changes. Returned that logic back.




2.It seems like the field validation is failing to work. In this case you’re supposed to get «Not enough funds» message. 0.104 is too much since `0.104 + fee` will result in total `amount > balance`. There is check against it.

I’ve tested the case with exactly same amount and fee that you have in the video. It seems to work fine for me. Could you, please, check if you have any local changes in `calculate()` function in `stakeFormEthereumActions` (this is where the check with balance happens) or maybe in compose logic on higher levels? Here’s the line: https://github.com/everstake/trezor-suite/pull/7/files#diff-d9ca54e1a6f65e4092fd0072f507d93168ede669cd875195707e2d7fbd5ee4f6L54


https://github.com/everstake/trezor-suite/assets/42133844/e1b801ff-f822-4635-a65d-7519729edc6f



If I emulate your case by returning in the mentioned above `calculate()` some unhandled error, then I can see the case when a user is stuck on the confirm modal. This happens because `composedTx` in `signTx()` is not of type final or undefined. So you can also check what you receive in as `composedTx` in `that signTx()`. Here’s the line: https://github.com/everstake/trezor-suite/blob/feat/stake-eth-action/packages/suite/src/hooks/wallet/useStakeEthForm.ts#L342

tomasklim commented:
I don't have any local changes.

https://github.com/everstake/trezor-suite/assets/33235762/ace70017-110c-4c27-8adf-9ff33eec4d72

The problem is that `formValues.selectedFee` is changed to `custom` and there is `predefinedLevels` with fee `custom` which is lower than `normal`

Probably added here https://github.com/everstake/trezor-suite/pull/7/files#diff-d9ca54e1a6f65e4092fd0072f507d93168ede669cd875195707e2d7fbd5ee4f6R169


<img width="347" alt="Screenshot 2024-01-20 at 19 26 52" src="https://github.com/everstake/trezor-suite/assets/33235762/e0f82068-8a41-4a52-aec9-c601a4ac8edc">


`errors` from `formState` is empty. However, `transactionInfo` from `composedLevels` with `custom` fee has error.
`transactionInfo` is used to show `Maximum fee` in the form. `errors` (`formIsValid`) for `Continue` button
<img width="445" alt="Screenshot 2024-01-20 at 19 35 07" src="https://github.com/everstake/trezor-suite/assets/33235762/77ac4b13-115c-45b6-a4f4-4a3f3b9bc49a">

StanCheb commented:
Is the behaviour the same with Holesky?

tomasklim commented:
I think the problem is not on Holesky as the gas price is `1`

StanCheb commented:
I’ve managed to test with real ETH and finally reproduced the problem. So, the problem was indeed in programmatic switching from `normal` to `custom` fee. It tries to switch to the nearest possible fee when there’s an error with `normal` fee. 

This doesn’t happen with devnet because `maxFee` never greater than `minFee` in `composeTransaction()` and thus the `custom` fee object never created. With real ETH it is created and switched to when `normal` fee calculations throw errors. 

I disabled switching to nearest available fee in `useStakeCompose` as well as forming custom fee in `stakeFormEthereumActions`. Left the code commented for future implementations of custom fees.

https://github.com/everstake/trezor-suite/pull/7/commits/ac6e548baa4396c020fab3aaea9a6db672b42315
